pattern.go

```diff
@@ -0,0 +1,227 @@
+package zero
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/wdvxdr1123/ZeroBot/message"
+)
+
+// PatternRule check if the message can be matched by the pattern
+func PatternRule(pattern *Pattern) Rule {
+	return func(ctx *Ctx) bool {
+		if len(ctx.Event.Message) == 0 {
+			return false
+		}
+		// copy messages
+		msgs := make([]message.Segment, 0, len(ctx.Event.Message))
+		msgs = append(msgs, ctx.Event.Message[0])
+		for i := 1; i < len(ctx.Event.Message); i++ {
+			if ctx.Event.Message[i-1].Type == "reply" && ctx.Event.Message[i].Type == "at" {
+				// [reply][at]
+				reply := ctx.GetMessage(ctx.Event.Message[i-1].Data["id"])
+				if reply.MessageID.ID() == 0 || reply.Sender == nil || reply.Sender.ID == 0 {
+					// failed to get history message
+					msgs = append(msgs, ctx.Event.Message[i])
+					continue
+				}
+				if strconv.FormatInt(reply.Sender.ID, 10) != ctx.Event.Message[i].Data["qq"] {
+					// @ other user in reply
+					msgs = append(msgs, ctx.Event.Message[i])
+				}
+			} else {
+				msgs = append(msgs, ctx.Event.Message[i])
+			}
+		}
```

Comment:
太啰嗦了，最后大部分都归结为`msgs = append(msgs, ctx.Event.Message[i])`，既然如此全部反过来判断更简洁。但我说实话这个判断真没必要，你这实际上是删除了@本人，不删除又何妨，删除了反而会让用户体验不一致。
```suggestion
		for i := 1; i < len(ctx.Event.Message); i++ {
			if ctx.Event.Message[i-1].Type == "reply" && ctx.Event.Message[i].Type == "at" {
				// [reply][at]
				reply := ctx.GetMessage(ctx.Event.Message[i-1].Data["id"])
				if reply.MessageID.ID() != 0 && reply.Sender != nil && reply.Sender.ID != 0 &&  strconv.FormatInt(reply.Sender.ID, 10) == ctx.Event.Message[i].Data["qq"] {
					continue
				}
			}
			msgs = append(msgs, ctx.Event.Message[i])
		}
```

Reply:
因为回复的时候那个@旧版客户端会自动加上，这个东西会干扰接下来的解析

Reply:
可以给出示例，如果这个没什么大用的话，直接在preprocess中删去。

Reply:
反正回复约等于@其他人，用户可以在Reply后面加一个Optional的At消除可能的匹配。如果真的匹配到了就让用户自己判断。

Reply:
> Reply后面加一个Optional的At消除可能的匹配

这样会不会有点麻烦，给Reply加个option默认过滤掉或者不过滤掉？

Reply:
也可以